Add ISA_dep_plugin for dependency checking and visualization

dallagi opened the pull request:
This plugin currently builds internally a dependency graph, renders it with graphviz and returns both a .dot file representing the graph and, if graphviz is installed, a visual representation of the graph.
Temporary data is stored in the report directory, and automatically deleted after a succesfull build.

ereshetova committed:
Merge pull request #17 from 01org/devel

Taking pkg alias info into use in CVE plugin

Commit:
checksec: Add explicit runtime dependency on binutils

Since checksec.sh functionality relies on readelf(1) utility,
which is part of binutils package, we should add runtime
dependencies on binutils explicitly.

Signed-off-by: Sergey Popovich <popovich_sergei@mail.ua>

ereshetova committed:
Making junit xml output work for every plugin

ereshetova committed:
Merge pull request #20 from 01org/devel

Making junit xml output work for every plugin

ereshetova committed:
Adding checks for executable stack to CFA plugin

ereshetova committed:
Merge pull request #21 from 01org/devel

Adding checks for executable stack to CFA plugin

ereshetova committed:
Adding checks for executables that don't drop groups while calling setuid/setgid to CFA plugin

ereshetova committed:
Merge pull request #22 from 01org/devel

 Adding checks for executables that don't drop groups properly

ereshetova committed:
Adding a check for CONFIG_X86_INTEL_MPX to KCA plugin

ereshetova committed:
Adding checks for mpx for executables in CFA plugin

ereshetova committed:
Merge pull request #23 from 01org/devel

Adding MPX-related checks

Commit:
isafw.bbclass: Use Base Package Name (BPN) instead of just Package Name (PN)

While we skipping analysesource task for most of native, nativesdk,
cross, crosssdk and other stuff we still have packages like glibc-initial
in our report.

This is because such initial packages are not belonging to any specific
target/image class and thus not subject to the analysesource task removal.

To address this issue use Base Package Name (BPN) since it just version of
Package Name (PN) with stripped suffixes like -initial, -native, -cross etc.

Signed-off-by: Sergey Popovich <popovich_sergei@mail.ua>

Commit:
isafw.bbclass: Add documentation strings describing common layer tasks

Two common and public tasks exposed by the class are analyse_sources
and analyse_sources_all. Both used to generate ISAFW reports against
given package(s) without building them.

Add descriptive text strings to the task definitions within class to
give more information on them to the user in the listtasks bitbake
task output.

  $ bitbake -c listtasks <target>
  do_analyse_sources             Produce ISAFW reports based ...
  do_analyse_sources_all         Produce ISAFW reports for all ...
  ...

TODO: find a way to hide class private tasks do_analysesource and
      do_process_reports.

Signed-off-by: Sergey Popovich <popovich_sergei@mail.ua>

Commit:
README.md: Add more Markdown styles to the text

Signed-off-by: Sergey Popovich <popovich_sergei@mail.ua>

ereshetova committed:
Additional fixes for xml output

dallagi committed:
Add ISA_dep_plugin for dependency checking and visualization

This plugin currently builds a dependency graph internally and
puts it in the log file.
Temporary data is stored in the report directory, and automatically
deleted after a succesfull build.

ereshetova committed:
Merge pull request #25 from 01org/devel

Additional fixes for xml output

ereshetova committed:
Merge pull request #24 from popovich-sergei/next

Next

ereshetova committed:
Fetch checksec instead of storing the sript

ereshetova committed:
Merge pull request #27 from 01org/devel

Fetch checksec instead of storing the sript

dallagi committed:
Render dependency graph

After the dependency graph has been internally created, generate a
.dot file representing the graph and call graphviz to render it.
Transitive reduction is applied to the graph definition.

dallagi committed:
Merge remote-tracking branch 'origin_old/pkg_dep' into pkg_dep

ereshetova committed:
Making plugins full reports produced only on request

ereshetova committed:
adding check of range for pkg list

ereshetova committed:
Adding machine name to report names, small cleanup

dallagi committed:
Generate run-time dependency graph

Dependencies are now saved as an object <dep, details>, as run-time
dependencies include details like package version.
Dependencies are filtered based on DEPENDENCIES_BLACKLIST.

A class has been used instead of a namedtuple because openembedded
monkey-patches the namedtuple module with an alternative one which
is not pickle-able.

dallagi committed:
Merge remote-tracking branch 'origin_old/master' into pkg_dep

dallagi committed:
Save graphs as svg instead of png

PNG graphs were often unreadable, especially in very large graphs.
While SVG files are not as convenient and easy to open, they are more
suitable for this purpose.

dallagi committed:
Add `repr` and `str` methods on `Dependency` class.

Those methods are useful when debugging.

ereshetova commented:
ok to test

kad commented:
retest this please

kad commented:
I would strongly recommend to rebase this PR, as there are too many "merge commits" present in change set. @dallagi please rebase your changes on top of origin/master, making sure that only your change commits are present in PR.

ereshetova commented:
@kad, this commit is for me to track progress of the work that @dallagi does, so there is no need to rebase it yet. This work is not ready to be merged to main repo for a while still.

dallagi committed:
Move graphs loading from temp into dedicated method

dallagi committed:
Generate partial dependency graphs for vulnerability reports

Add the possibility to generate partial dependency graphs.

Dependency graphs are now color-coded: vulnerable packages
are represented by red nodes, while packages that depend on vulnerable
packages are represented by yellow nodes.

Edges in dependency graphs now contain info (e.g., pkg version) when
those are provided.

dallagi committed:
Add hooks for complete/partial vulnerability graph generation

dallagi committed:
Add hooks for build-time vulnerability graph generation

dallagi committed:
Add cleanup method to isafw

dallagi committed:
Add method to get a list of all the binaries generated by some source packages

Mapping of source packages to the binary packages they generate is built
during the package-processing phase, and it is stored in a temporary file.
